### DISCUSSION TITLE
Is it possible to detect collisions between objects(meshes) in the "Editor mode"?

### DISCUSSION THREAD

**Discussion by interceptor47** [o3de/o3de#10478] [Help Wanted]
Say, I have two objects/meshes - a cube and a sphere. In the **Editor mode**, if I click on the cube (using a mouse) and drag it over the sphere where it overlaps it, is it possible to detect collisions (overlapping/intersecting events are generated) between the two objects?

Essentially, is collision detection possible in the **Editor mode**  and not in the Play mode, where I know it's possible.

Any pointers would be appreciated. Thanks in advance.

**Comment by AMZN-alexpete** [comment]
@SergeyAMZN  may know more about this

I think it is possible using the Editor Physics scene instead of the default one. 

You might be able to add your own editor physics trigger shapes in the editor version of your components and test for collisions - but you need to make sure they are added to the PhysX editor scene `AzPhysics::EditorPhysicsSceneName;` https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsTypes.h#L30
I'm pretty sure that scene is active in the editor when not in game mode: https://github.com/o3de/o3de/blob/development/Gems/PhysX/Code/Editor/Source/Components/EditorSystemComponent.cpp#L133

**Reply by SergeyAMZN** [reply]
Yes, we do have Editor physics scene with all colliders added into it. 

Make sure the entities you want to overlap with do have a collider component though!
It could be one of these:
* HeightfieldColliderComponent
* ColliderComponent
* ShapeColliderComponent
* WhiteboxColliderComponent

Check https://github.com/o3de/o3de/blob/development/Gems/RecastNavigation/Code/Source/Misc/RecastNavigationPhysXProviderComponentController.cpp for an example: `RecastNavigationPhysXProviderComponentController::CollectCollidersWithinVolume`

For the sphere overlap it will be 
`AzPhysics::OverlapRequestHelpers::CreateSphereOverlapRequest` and then             
```
AzPhysics::SceneHandle sceneHandle = sceneInterface->GetSceneHandle(AzPhysics::EditorPhysicsSceneName);
sceneInterface->QueryScene(sceneHandle, &request);
```

**Reply by interceptor47** [reply]
Thanks for the reply. Will try out the suggestions.

**Reply by interceptor47** [reply]
Still new to O3DE and how to modify/extend it. Had a couple of questions on this.

**Do I have to replace the default editor and create a new one to enable this functionality? And if yes, do I have to do this by creating a new (editor?) gem, which would have the functionality as shown in the example?**

Also, is there any documentation to understand the code-base? Maybe a high level walkthrough of the overall flow - e.g. a sequence diagram that details which classes are called when the engine starts?

[Edit1] 
For context why I need this — I’m creating an electromechanical simulator application, where the requirement is to know *before* the play/simulation begins, if two objects are “connected to each other”. e.g. a plug object connected to a socket → a wire connected to the plug on one end and to an electrical appliance at the other end → A sensor attached (connected) to the appliance to measure its temperature. And, so on. Essentially, creating/designing an electromechanical circuit – in the Editor – and then simulating/playing it.  

So, in O3DE, when the scene loads I'd select an object in the Editor (say, a plug) using the mouse and drag it over to the other object (say, a socket) to "connect" to it . And when these two objects overlap with each other (connect to each other) one of the object would change color. 

Once the circuit is "designed" in the Editor, click on simulate/play to observe their behavior (e.g. the amount of current flowing through the wire, or the temperature of the appliance etc.). 

[Edit2] 
From what I have understood, I'd need to enable collisions on the objects (using collider component) and then query the scene to get a list of overlapping objects using  _sceneInterface->QueryScene(sceneHandle, &request);_.  Once I do that, I also need to visually display in the Editor which objects are overlapping. E.g. in the above example, when the user clicks-n-drags the plug to the socket and they overlapp, I'd want to change their color, indicating to the user that these objects are "connected/wired" to each other. **How do I display that information in the Editor? Or, how to I "write visual information" to the Editor?**

Thanks.
